common/src/test/java/org/cloudfoundry/identity/uaa/util/UaaUrlUtilsTest.java

```diff
@@ -35,6 +39,74 @@ public void tearDown() throws Exception {
     }
 
     @Test
+    public void testEncodedUrl() throws Exception {
+        String url = "https://example.com/dashboard/?appGuid=app-guid&ace_config=%7B%22orgGuid%22%3A%22org-guid%22%2C%22spaceGuid%22%3A%22space-guid%22%2C%22appGuid%22%3A%22app-guid%22%2C%22redirect%22%3A%22https%3A%2F%2Fexample.com%2F%22%7D";
+        String extraValueEncoded = "this%20is%20my%25special%20string%20that%20has%26some%3Dspecial%20values";
+        String extraValueDecoded = URLDecoder.decode(extraValueEncoded, "UTF-8");
+        UriComponentsBuilder template = uaaURLUtils.parseAndDecodeUrl(url);
+        uriEquals(new URI(url), new URI(template.build().encode().toUriString()));
+        url += "&param1="+extraValueEncoded;
+        template.queryParam("param1", extraValueDecoded);
+        uriEquals(new URI(url), new URI(template.build().encode().toUriString()));
+        url += "&param2="+extraValueEncoded;
+        template.queryParam("param2", "{param2}");
+        uriEquals(new URI(url), new URI(template.build().expand(extraValueDecoded).encode().toUriString()));
+    }
+
+    public static void uriEquals(URI uri1, URI uri2) {
```

Comment:
the Spring UriComponentsBuilder uses a very strict encoding scheme. It goes by the letter for the HTTP specification of what is allowed for different components of the URI.

However, java.net.URLEncoder is a superset of the UriComponentsBuilder encoding, thus, after we encode the string using UriComponentsBuilder, it no longer looks the same. Comparison can only be done if we use decoded values.